### PR TITLE
Recurrent: reshape input to boost the performance

### PR DESCRIPTION
**Pull request by JerryYanWan**
fastmode test

## What changes were proposed in this pull request?
In Recurrent layer, previously use Tensor.select(2, i) to form a 2D matrix, resulting in a discontiguous storage. Add a transpose layer ahead and after the model to boost the performance.

## How was this patch tested?

test comparison between the original recurrent layer and boosted layer. 
They should yield the same outputs.
```
    val batchSize = 10
    val nWords = 5
    val hiddenSize = 4
    val inputSize = 5
    val outputSize = 5
    val bpttTruncate = 3
```
The boosted recurrent layer output:
```
1-th loss = 1.7208151374781946
2-th loss = 1.6437483026807242
3-th loss = 1.588589862337444
4-th loss = 1.546502647475069
5-th loss = 1.5122343688179611
6-th loss = 1.4820277319857593
7-th loss = 1.4532184745823775
8-th loss = 1.423996271305294
9-th loss = 1.393202500769242
10-th loss = 1.3601207835231857
11-th loss = 1.3243031658388933
12-th loss = 1.2854901773062697
13-th loss = 1.2436207891776232
14-th loss = 1.1988841429635386
15-th loss = 1.1517747080464198
16-th loss = 1.1031160939960707
17-th loss = 1.0539955861618664
18-th loss = 1.0055591811744848
19-th loss = 0.9587237255737998
20-th loss = 0.9139874263089901
21-th loss = 0.8714513132386527
22-th loss = 0.8309683094621219
23-th loss = 0.7922852310298951
24-th loss = 0.7551259365838042
25-th loss = 0.71922366386225
26-th loss = 0.6843226960866249
27-th loss = 0.6501750701749818
28-th loss = 0.6165679246376932
29-th loss = 0.5834084797893359
30-th loss = 0.5508311175279348
31-th loss = 0.5191993675987441
32-th loss = 0.4889363857214196
33-th loss = 0.46034216895242297
34-th loss = 0.4335482614955519
35-th loss = 0.40856704821033113
36-th loss = 0.38535108296623966
37-th loss = 0.3638288337901121
38-th loss = 0.34391885414011847
39-th loss = 0.32553310008330955
40-th loss = 0.3085772198541273
41-th loss = 0.29295144097855785
42-th loss = 0.27855289090984864
43-th loss = 0.2652787376831657
44-th loss = 0.2530292753685399
45-th loss = 0.2417103143777617
46-th loss = 0.23123467134057182
47-th loss = 0.22152282538857349
48-th loss = 0.21250296062898827
49-th loss = 0.20411061436291272
50-th loss = 0.19628811821965705
```

The original recurrent layer output:
```
1-th loss = 1.7208151374781946
2-th loss = 1.6437483026807242
3-th loss = 1.588589862337444
4-th loss = 1.546502647475069
5-th loss = 1.5122343688179611
6-th loss = 1.4820277319857593
7-th loss = 1.4532184745823775
8-th loss = 1.423996271305294
9-th loss = 1.393202500769242
10-th loss = 1.3601207835231857
11-th loss = 1.3243031658388933
12-th loss = 1.2854901773062697
13-th loss = 1.2436207891776232
14-th loss = 1.1988841429635386
15-th loss = 1.1517747080464198
16-th loss = 1.1031160939960707
17-th loss = 1.0539955861618664
18-th loss = 1.0055591811744848
19-th loss = 0.9587237255737998
20-th loss = 0.9139874263089901
21-th loss = 0.8714513132386527
22-th loss = 0.8309683094621219
23-th loss = 0.7922852310298951
24-th loss = 0.7551259365838042
25-th loss = 0.71922366386225
26-th loss = 0.6843226960866249
27-th loss = 0.6501750701749818
28-th loss = 0.6165679246376932
29-th loss = 0.5834084797893359
30-th loss = 0.5508311175279348
31-th loss = 0.5191993675987441
32-th loss = 0.4889363857214196
33-th loss = 0.46034216895242297
34-th loss = 0.4335482614955519
35-th loss = 0.40856704821033113
36-th loss = 0.38535108296623966
37-th loss = 0.3638288337901121
38-th loss = 0.34391885414011847
39-th loss = 0.32553310008330955
40-th loss = 0.3085772198541273
41-th loss = 0.29295144097855785
42-th loss = 0.27855289090984864
43-th loss = 0.2652787376831657
44-th loss = 0.2530292753685399
45-th loss = 0.2417103143777617
46-th loss = 0.23123467134057182
47-th loss = 0.22152282538857349
48-th loss = 0.21250296062898827
49-th loss = 0.20411061436291272
50-th loss = 0.19628811821965705
```
The speed for boosted recurrent should be faster than the lecayMode:
```
    val batchSize = 50
    val nWords = 20
    val hiddenSize = 40
    val inputSize = 1000
    val outputSize = 1000
    val bpttTruncate = 5
    val nEpoch = 50
```
Note that the testing data might variate given the data randomness.

|  (second)        | forward           | backward  |
| ---------------- |:-------------:| -----:|
| Boosted        | 0.01561445816 | 0.06007293678  |
| Original | 0.01116085474 |  0.1658622401  |

The boosting mode might not outperform original model for the forward because the time for Tensor.resize.copy operations compensate the boosting time for Tensor.select. However, it diminishes the backward time noticeably.

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 1]
```diff
@@ -21,18 +21,71 @@ import com.intel.analytics.bigdl.Module
...
+   *
+   * For the first scenario, the input tensor might be transpose before recurrent
+   * layer. Thus, the legacyMode should be set to be true so as to select the
+   * correct dimension for recurrence.
+   *
+   * For the second one, the input tensor has not been transpose, and user can
+   * define legacyMode to be false to select correct time dimension for recurrence.
+   * The drawback of this mode is that the storage in the output Tensor will be
+   * discontiguous, which will slow the processing speed.
+   */
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: Update comments

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 2]
```diff
@@ -21,18 +21,71 @@ import com.intel.analytics.bigdl.Module
...
-  var (batchSize, times) = (0, 0)
+  var (batchSize, times, nDim) = (0, 0, 0)
+
+  /**
+   * batchDim should be either 1 or 2 according to timeDim
+   * hiddenSelect is always equal to 1
+   * outputSelect is always equal to timeDim
+   */
+
+  val batchDim = 3 - timeDim
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: what if timeDim is set to, say, 4?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 3]
```diff
@@ -45,19 +98,26 @@ class Recurrent[T : ClassTag] (
...
     transform = modules(1)
 
-    batchSize = input.size(1)
-    times = input.size(2)
+    batchSize = input.size(batchDim)
+    times = input.size(timeDim)
+
+    hidden.resize(Array(times + 1, batchSize, hiddenSize))
+
+    if (timeDim == 2 && !legacyMode) {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: if (dataSelect != timeDim)

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 4]
```diff
@@ -45,19 +98,26 @@ class Recurrent[T : ClassTag] (
...
     var i = 1
     while (i <= times) {
-      val curInput = T(input.select(2, i), hidden.select(2, i))
+      val curInput = T(fInput.select(dataSelect, i), hidden.select(hiddenSelect, i))
       val currentOutput = module.updateOutput(curInput)
       transform.updateOutput(currentOutput)
-      output.select(2, i).copy(transform.output.toTensor)
-      hidden.select(2, i + 1).copy(transform.output.toTensor)
+      output.select(outputSelect, i).copy(transform.output.toTensor)
+      hidden.select(hiddenSelect, i + 1).copy(transform.output.toTensor)
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: Maybe we can change hidden to Array[Tensor](times + 1), and then we can avoid the copy here: hidden(i+1) = transform.output

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 5]
```diff
@@ -86,18 +148,25 @@ class Recurrent[T : ClassTag] (
...
+    } else {
+      fGradOutput = gradOutput
+    }
+
     var i = times
     while (i >= 1) {
       transform.output.toTensor
-        .copy(hidden.select(2, i + 1))
-      var deltaHidden = transform.updateGradInput(hidden.select(2, i), gradOutput.select(2, i))
+        .copy(hidden.select(hiddenSelect, i + 1))
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: why not: `transform.output = hidden.select(hiddenSelect, i + 1)`?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 6]
```diff
@@ -86,18 +148,25 @@ class Recurrent[T : ClassTag] (
   override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
     gradInput.resizeAs(input)
 
+    if (timeDim == 2 && !legacyMode) {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: if (dataSelect != timeDim)?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 7]
```diff
@@ -67,14 +127,16 @@ class Recurrent[T : ClassTag] (
                                  scale: Double = 1.0): Unit = {
     var i = times
     while (i >= 1) {
-      transform.output = hidden.select(2, i + 1)
-      var deltaHidden = transform.updateGradInput(hidden.select(2, i), gradOutput.select(2, i))
+      transform.output = hidden.select(hiddenSelect, i + 1)
+      var deltaHidden = transform.updateGradInput(
+        hidden.select(hiddenSelect, i), fGradOutput.select(dataSelect, i))
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: Shall we store deltaHidden in a temporary tensor?

**Comment by jason-dai** [comment]
Do you have updated benchmarking data?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala (hunk 2 above)
Comment (JerryYanWan) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: The input should be a 3D tensor, so it is not possible for the timeDim set to 4.
I have added a requirement for this argument.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala (hunk 3 above)
Comment (JerryYanWan) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: revised

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala (hunk 6 above)
Comment (JerryYanWan) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: revised

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala (hunk 1 above)
Comment (JerryYanWan) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: updated

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 8]
```diff
@@ -21,18 +21,69 @@ import com.intel.analytics.bigdl.Module
...
+  /**
+   * The Recurrent will always go along the first dimension of the Tensor given that
+   * the legacyMode is set to be false in order to boost the performance.
+   *
+   * If the timeDim is 2, this layer will transpose the first two dimensions of the input.
+   * Otherwise, the input will remain unchanged.
+   *
+   * The legacyMode is set solely for unit test to compare the output of this layer. To make
+   * sure that the outputs will be the same regardless of transposition of input.
+   */
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: Update comment

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 9]
```diff
@@ -21,18 +21,69 @@ import com.intel.analytics.bigdl.Module
...
+
+  require(timeDim < 3,
+    "In Recurrent: the timeDim should be less than three," +
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: the timeDim should be 1 or 2

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 10]
```diff
@@ -45,19 +96,30 @@ class Recurrent[T : ClassTag] (
...
-    batchSize = input.size(1)
-    times = input.size(2)
+    batchSize = input.size(batchDim)
+    times = input.size(timeDim)
 
-    output.resize(Array(batchSize, times, hiddenSize))
-    hidden.resize(Array(batchSize, times + 1, hiddenSize))
+    hidden.resize(Array(times + 1, batchSize, hiddenSize))
+
+    if (dataSelect != timeDim && !legacyMode) {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: do we still need legacyMode here?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 11]
```diff
@@ -86,18 +150,30 @@ class Recurrent[T : ClassTag] (
   override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
     gradInput.resizeAs(input)
 
+    if (dataSelect != timeDim && !legacyMode) {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: do we still need legacyMode here?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 12]
```diff
@@ -21,18 +21,69 @@ import com.intel.analytics.bigdl.Module
...
+  var dataSelect = 1
+  val outputSelect = timeDim
+  var fInput: Tensor[T] = _
+  var fGradOutput: Tensor[T] = _
+
+  /**
+   * The legacyMode is set to perform an Recurrence
+   * with non-contiguous Tensor
+   */
+  def setLegacyMode(): Unit = {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: make it private and document that it is for test only.

File: dl/src/test/scala/com/intel/analytics/bigdl/nn/RecurrentSpec.scala [hunk 13]
```diff
@@ -21,12 +21,12 @@ import com.intel.analytics.bigdl.optim.SGD
 import com.intel.analytics.bigdl.tensor.Tensor
 import com.intel.analytics.bigdl.utils.RandomGenerator.RNG
 import com.intel.analytics.bigdl.utils.T
-import org.scalatest.{FlatSpec, Matchers}
+import org.scalatest.{FlatSpec, Matchers, PrivateMethodTester}
```
Comment (jason-dai) [inline dl/src/test/scala/com/intel/analytics/bigdl/nn/RecurrentSpec.scala]: Do we still need PrivateMethodTester?

Comment (JerryYanWan) [inline dl/src/test/scala/com/intel/analytics/bigdl/nn/RecurrentSpec.scala]: removed

**Comment by jason-dai** [comment]
Please update the comments and the description of the PR; otherwise LGTM.

**Pull request closed without merging**